File: ios/gardenapp/WidgetDataSyncModule.m
//
//  WidgetDataSyncModule.m
//  Bridge file for WidgetDataSyncModule
//

#import <React/RCTBridgeModule.h>
#import <React/RCTLog.h>

@interface RCT_EXTERN_MODULE(WidgetDataSync, NSObject)

RCT_EXTERN_METHOD(syncToAppGroup:(NSString *)data
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(syncPlannerToAppGroup:(NSString *)data
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(syncFinanceToAppGroup:(NSString *)data
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

File: ios/gardenapp/WidgetDataSyncModule.swift
//
//  WidgetDataSyncModule.swift
//  Native module to sync widget data to App Groups
//

import Foundation
import React
import WidgetKit

@objc(WidgetDataSync)
class WidgetDataSync: NSObject {
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  @objc
  func syncToAppGroup(_ data: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    let appGroupIdentifier = "group.com.sriharigurugubelli.gardenapp"
    
    guard let sharedDefaults = UserDefaults(suiteName: appGroupIdentifier) else {
      rejecter("APP_GROUP_ERROR", "Failed to access App Group: \(appGroupIdentifier)", nil)
      return
    }
    
    sharedDefaults.set(data, forKey: "widget_data")
    sharedDefaults.synchronize()
    
    // Reload widget timelines
    if #available(iOS 14.0, *) {
      WidgetCenter.shared.reloadTimelines(ofKind: "HabitsWidget")
    }
    
    resolver(true)
  }
  
  @objc
  func syncPlannerToAppGroup(_ data: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    let appGroupIdentifier = "group.com.sriharigurugubelli.gardenapp"
    
    guard let sharedDefaults = UserDefaults(suiteName: appGroupIdentifier) else {
      rejecter("APP_GROUP_ERROR", "Failed to access App Group: \(appGroupIdentifier)", nil)
      return
    }
    
    sharedDefaults.set(data, forKey: "planner_widget_data")
    sharedDefaults.synchronize()
    
    // Reload widget timelines
    if #available(iOS 14.0, *) {
      WidgetCenter.shared.reloadTimelines(ofKind: "PlannerWidget")
    }
    
    resolver(true)
  }
}

